Added: Email integration workflow (#9)
hotwax/oms-documentation: Journeys/Email integration/Promise date change communication.md

# Integration guidelines for OMS and Marketing Automation Platforms: Promise Date Change Communication 

## Overview
This document outlines the steps to ensure accurate promise date change communication from the Order Management System (OMS) to the Marketing Automation Platform. This integration enhances the customer experience and reduces manual efforts.

### Step 1: Identify Promise Date Changes

> The bulkSendPromiseDateChange service in OMS identifies all the sales order items whose promise dates have changed in throughout the day. The frequency of this service is configurable as per requirement, but it is recommended to set it once a day.

Reviewer: Check grammar

Author: Made the changes. Removed 'in' before throughout

# Integration guidelines for OMS and Marketing Automation Platforms: Promise Date Change Communication 

## Overview
This document outlines the steps to ensure accurate promise date change communication from the Order Management System (OMS) to the Marketing Automation Platform. This integration enhances the customer experience and reduces manual efforts.

### Step 1: Identify Promise Date Changes

> The BulkSendPromiseDateChange service in OMS identifies all the sales order items whose promise dates have changed in the last 30 minutes. The frequency of this service is configurable as per requirement, but it is recommended to set it to 30 minutes.

Reviewer: Where did we get this recommended time? 30 min seems a little low, I think it should be closer to once a day.

Author: Yes, you are right. It is scheduled for once in a day, the service which changes the promise dates on PDP is scheduled for every half an hour. Confirmed it from Deepak sir. 

Changing it now

#### Service Details:
```
<service name="bulkSendPromiseDateChangedNotification" engine="java"
             location="co.hotwax.customerservice.order.CsrOrderServices" invoke="bulkSendPromiseDateChangedNotification" auth="true" transaction-timeout="3600">
        <description>Bulk service to send promised date changed notification. Default send promised date changed on the last day.</description>
        <attribute name="fromDate" type="Timestamp" mode="IN" optional="true"/>
        <attribute name="toDate" type="Timestamp" mode="IN" optional="true"/>
</service>
```
| Parameter | Description | Required |
|-----------|-------------|----------|
| `fromDate` | The start time for the service to read the change | No |

> | `toDate` | The end time for the service to read the change| No |

Reviewer: ```suggestion
| `fromDate` | The date and time from which to find order item changes | No |
| `toDate` | The date and time from which to find order item changes till | No |
```

Make sure that its from and to of the order item change and not the order item itself

Author: sure, making the change accordingly

### Step 2: Prepare for Email Notification

> The BulkSendPromiseDateChange service picks up all the updated sales order items. This service processes the data and prepares it for the email notification by bringing all the sales order items whose promise dates have changed between `fromDate` and `toDate`.

Reviewer: ```suggestion
The BulkSendPromiseDateChange service identifies all the updated sales order items whose promise dates have changed between the `fromDate` and `toDate` and gathers order context needed to hand off to the Marketing Automation Platform. 
```

### Step 3: Call `sendDeliveryDateChangedNotification` service
After bringing all the updated items in response, the BulkSendPromiseDateChange service calls the `sendDeliveryDateChangedNotification` service for each item.

#### Service Details:
```
<service name="sendDeliveryDateChangedNotification" engine="java" require-new-transaction="true" max-retry="3"
             location="co.hotwax.warehouse.shipping.WarehouseServices" invoke="sendDeliveryDateChangedNotification">
        <description>Send an order item Delivery Date Changed notification</description>
        <implements service="orderNotificationInterface"/>
        <attribute name="orderId" type="String" mode="IN" optional="true"/>
        <attribute name="orderItemSeqId" type="String" mode="IN" optional="true"/>
</service>
```
| Parameter | Description | Required |
|-----------|-------------|----------|
| `orderId` | The ID of the order in Shopify | No |

> | `orderItemSeqId` | The ID of the sequence of an item in the order | No |

Reviewer: how can both of these be optional? one of these has to be required, and I think both are.

Author: Confirmed again with Nameet sir, orderID is required

Reviewer: make a ticket for this in the OMS git

### Step 4: Customize Email Notification

> The `sendDeliveryDateChangedNotification` service checks the configuration for the email integration platform and picks the appropriate API template. The service then takes input parameters  `orderId` and `orderItemSequenceID` to update all the sales order items in the API template with changed promised dates.

Reviewer: ```suggestion
The `sendDeliveryDateChangedNotification` service identifies the configured email integration platform and picks the appropriate API template. The service then uses the input parameters  `orderId` and `orderItemSequenceID` to add all the sales order items with changed promise dates in the email template being pushed to the Marketing Automation Platform.
```

What do you mean by "update all the sales order items..."?
What if there is more than 1 email platform configured?

Author: Understood, I meant similar to your version but it was worded incorrectly. Thank you for correcting.

HotWax Commerce has ready integration with Klaviyo, a marketing automation platform. Here's a sample JSON file that is shared with Klaviyo's API for each order:
```
{
  "brandName": "",
  "DISCOUNT_CODE": "",
  "DISCOUNT_VALUE": "",
  "ORDERNO": "",
  "DISCOUNT_CODE": "",
  "FIRSTNAME": "",
  "ITEMLIST_XML": [
    {
      "itemId": "",
      "status": "",
      "description": "",
      "quantity": "",
      "color": "",
      "price": "",
      "itemImage": "",
      "itemUrl": "",

>       "backOrder": "",

Reviewer: is there no parameter for pre-order?

Author: There is a parameter for pre-order as well, this template is configured as per client requirement, if needed it can be changed to preorder.

Reviewer: Discuss this with Deepak, there should be one template that works for everyone because the data is the same. Make a ticket in the OMS git accordingly and update this document with what the fields should be not what they are today.

Reviewer 2: This parameters is not in use, earlier it was used to show some tag at item level on email.
As per current implementation value of this field is empty.
We can ignore or remove this from documentation

Reviewer 2: > Discuss this with Deepak, there should be one template that works for everyone because the data is the same. Make a ticket in the OMS git accordingly and update this document with what the fields should be not what they are today.

This is the single template used for both preorder/backorder

Reviewer: How do we know which one it is? Is this a Y/N flag?

Reviewer 2: Right now we are not sending anything on backorder field

Reviewer 2: > implementation

@dt2patel

Reviewer 2: > This parameters is not in use, earlier it was used to show some tag at item level on email. As per current implementation value of this field is empty. We can ignore or remove this from documentation

@dt2patel

Author: Hey @dt2patel 
As discussed with @dixitdeepak sir, we currently don't have a ready json format to identify the type (pre-order/backorder) of the order item. We can create one if necessary, but considering there is no use case for order item type yet, we have not done so. He has advised to exclude the field for the time being; it can be added later if the format changes. 

My suggestion: We should use the Json format as is for the scope of this document, with the note that how it should be used to tell the order item type. Later, when we change the Json format, we can sync the changes here.

      "deliveryDate": "",
      "clearance": "",
      "itemName": "",
      "size": "",
      "itemUri": []
    }
  ]
}
```

> To update the delivery date, the order item is identified by the `itemID` and the `deliveryDate` is updated.

Reviewer: Not sure what this means, can you explain this?

Author: I have removed this sentence, when I read this now it is confusingly worded and the similar information is already conveyed. 

Explanation: Order is identified with the itemID and the 'deliveryDate' value is used as the new promise date.

### Step 5: Deliver Email Notification

> The service calls dedicated email platform's API, sends the details to the email integration platform and the email will be delivered to the customers. 

Reviewer: ```suggestion
The service calls the email platform's API and sends the details to the email integration platform which will then further deliver it to the customer.
```
